gridstatus/miso_api.py

```diff
@@ -0,0 +1,321 @@
+        # Interval format is hh:mm at the start of the interval
+        interval = date.floor("5min").strftime("%H:%M")
+        date_str = date.strftime("%Y-%m-%d")
+        version = EX_ANTE
+
+        url = (
+            f"{BASE_PRICING_URL}/real-time/{date_str}/lmp-{version}?interval={interval}"
+        )
+
+        data_list = self._get(url, verbose=verbose)
+
+        return data_list
+
+    @support_date_range(frequency="5_MIN", convert_to_dataframe=False)
+    def _get_lmp_real_time_5_min_ex_post(
+        self,
+        date,
+        end=None,
+        prelim_or_final: str = PRELIMINARY_STRING,
+        verbose=False,
+    ):
+        # Interval format is hh:mm at the start of the interval
+        interval = date.floor("5min").strftime("%H:%M")
+        date_str = date.strftime("%Y-%m-%d")
+        version = EX_POST
+        resolution = FIVE_MINUTE_RESOLUTION
+
+        url = f"{BASE_PRICING_URL}/real-time/{date_str}/lmp-{version}?interval={interval}&preliminaryFinal={prelim_or_final}&timeResolution={resolution}"  # noqa
+
+        data_list = self._get(url, verbose=verbose)
+
+        return data_list
+
+    def process_pricing_data(self, data_list: List[Dict], market: Markets):
+        df = pd.DataFrame(data_list)
+
+        # Split timeInterval dict into separate columns for 'start' and 'end'
+        df = pd.concat(
+            [
+                df["timeInterval"].apply(pd.Series)[["start", "end"]],
+                df.drop(columns=["timeInterval"]),
+            ],
+            axis=1,
+        )
+
+        df["Interval Start"] = pd.to_datetime(df["start"], utc=True).dt.tz_convert(
+            self.default_timezone,
+        )
+        df["Interval End"] = pd.to_datetime(df["end"], utc=True).dt.tz_convert(
+            self.default_timezone,
+        )
+
+        node_to_type_mapping = (
+            MISO()
+            ._get_node_to_type_mapping()
+            .set_index("Node")["Location Type"]
+            .to_dict()
+        )
+
+        df["Location Type"] = df["node"].map(node_to_type_mapping)
+        df["Market"] = market.value
+
+        df = df.rename(
+            columns={
+                "node": "Location",
+                "lmp": "LMP",
+                "mcc": "Congestion",
+                "mec": "Energy",
+                "mlc": "Loss",
+            },
+        )
+
+        # Column ordering
+        df = df[
+            [
+                "Interval Start",
+                "Interval End",
+                "Market",
+                "Location",
+                "Location Type",
+                "LMP",
+                "Energy",
+                "Congestion",
+                "Loss",
+            ]
+        ]
+
+        return df
+
+    def _get(self, url, verbose=False, max_retries=3):
+        headers = self._headers()
+        data_list = []
+
+        if verbose:
+            logger.info(f"Getting data from {url}")
+
+        response = requests.get(url, headers=headers, verify=False)
```

Comment:
Unfortunately, this doesn't work unless `verify=False`. The error is:

```
(Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)')))
```

I've tried using `verify=certifi.where()` which fails with the same error. We don't want to skip validation so I'm looking into other solutions.

Reply:
Running a test against the server shows an incomplete chain of certificates. It seems that a browser is able to find those missing certificates, but requests does not.

https://www.ssllabs.com/ssltest/analyze.html?d=apim.misoenergy.org
https://stackoverflow.com/a/66111417/5755357

<img width="1019" alt="Screenshot 2024-12-01 at 20 25 37" src="https://github.com/user-attachments/assets/4803caf9-8ae9-4b4f-946c-ba9c946b0727">